sirepo/simulation_db.py

```diff
@@ -151,8 +145,16 @@ def delete_user(uid):
     pkio.unchecked_remove(user_path(uid=uid))
 
 def examples(app):
+    d = None
+    try:
+        d = sirepo.sim_data.get_class(app).resource_path(_EXAMPLE_DIR)
+    except Exception as e:
+        if pkio.exception_is_not_found(e):
+            # Not all sim types have examples (ex jupyterhublogin)
+            return []
+        raise
     files = pkio.walk_tree(
```

Comment:
Did you not understand my comment?

Reply:
I did - I had responded here https://github.com/radiasoft/sirepo/pull/3747#issuecomment-899726777.

Thinking about this some more I think we need to decide what the semantics are because, as you say, they are inconsistent. Making examples augment is only making it more inconsistent. The concept of `sim_data.resource_path` doesn't make any sense if some resources have a single path (ex a jinja file) whereas other resources have multiple paths (ex example files).

Here's what I think should be done:
- pkresource should be changed so it doesn't do *any* overriding. We should make new method(s) that return a list of all matching filenames in all searched packages. We never stop once we've found one match. This is a change in behavior so we can probably just leave the old `filename` method and make new method(s) for Sirepo to use with new behavior. 
- Sirepo will then decide what makes sense for certain files. Ex, in the case of example it makes sense to augment (a sim can have any number of examples). In the case of say `landing.html` it doesn't make sense so if more than one file is returned then Sirepo will have to decide what it wants to do. pkresource can provide a guarantee that files are returned in the same order as the supplied packages so Sirepo can just use the 1st file returned if it wants the current override behavior. 
- All calls to `sim_data.resource_path` will be removed in favor of a call like `sim_data.example_paths`. The different function calls will know how to handle the files (whether or not multiple files makes sense or not). 

Thoughts?

Reply:
> I don't understand why this would be needed and more importantly how the caller would know what to do with the list of files, when it only wants one.

My thinking was that pkresource shouldn't be concerned with knowing when to override or when to get all files. Sirepo really knows this so pkresource should always return all matching files in all supplied packages and let Sirepo decide. For example, when we get examples we use all. In contrast to static files we would pick just one.  pkresource could guarantee that the files are returned in order of supplied packages so for static files sirepo would pick just the first file.

> glob_files(EXAMPLE_DIR/*.json) would work

Almost, except that it needs to prepend `template/<sim_type>` to the `EXAMPLE_DIR`

I think I've found a simple solution. Make `sim_data` be the only one that knows about `template` and make `sim_data.resource_path` always return a path to a file instead of a relative path to the resource dir.

e8fedc0df

Reply:
Unfortunately, I disagree. The problem we have is: map an abstract name to an actual path. The mapping in Sirepo's (and I argue almost all other cases) is one to one, that is, there is one and only one path for a name. 

Consider the problem of executing a command in Bash, you don't ever get a choice. Bash executes the first command it finds in $PATH. The same is true of the loader and $LD_LIBRARY_PATH. 

I am having a hard time thinking of a case where you want multiple values for a single abstract name in Sirepo. Do you have one? 

There are times in BOP that you want to know the "next one in the path", which is tricky, because you need to know where you are in the path (the concept iterates). This is handled by the BOP loader for class inheritance. We don't have this case in Sirepo or Pykern.

> Almost, except that it needs to prepend template/<sim_type> to the EXAMPLE_DIR

This is a detail that can be handled in this particular case, because it is very specific to this case (creating a URL, not an actual file). From the path, you get basename, and then prepend `template/<sim_type>`. This doesn't change the abstraction name to actual path mapping: it's still one to one.

Reply:
> Consider the problem of executing a command in Bash, you don't ever get a choice. Bash executes the first command it finds in $PATH. The same is true of the loader and $LD_LIBRARY_PATH.

I'm not sure I like that they do that:) I think tools like pyenv had to be created in part because of that behavior of `$PATH`.

I've had many run ins with `$LD_LIBRARY_PATH` where I wish it didn't pick the library for me.

Recently I was trying to compile emacs from source. The linker kept picking the wrong zlib (one with xcode not the one I installed). Supplying all sorts of args to impact `$LD_LIBRARY_PATH` seemed to have no effect. If it was designed where I had to decide which one to use I wouldn't have had this problem (although I'm sure the behavior would lead to other problems I don't see...).

> I am having a hard time thinking of a case where you want multiple values for a single abstract name in Sirepo. Do you have one?

In the [original issue](https://github.com/radiasoft/sirepo/issues/3732) you say, `private cloud customers have custom home pages`. My assumption is that would be implemented by overriding `landing.html`. This would mean we have two values for the single `static/html/landing.html` abstract name (one in Sirepo and one in the customer's repo). Another thought would be customers could override pages like forbidden.html so they could provide information specific to their organization (contact it@customer.com not support@radiasoft.net)

> This is a detail that can be handled in this particular case, because it is very specific to this case

Make sense, thanks. The change I linked to above obviates the need for prepending and instead keeps all of the details in `sim_data` (e8fedc0). Thoughts?